### ISSUE TITLE
WebTransport readable is never done

### ISSUE THREAD

**Issue by guest271314** [aiortc/aioquic#217]
Navigate to https://quic.aiortc.org/ open `console` and run the following two variations of the same code

```
(async () => {
  try {
    const wt = new WebTransport('https://quic.aiortc.org/wt');
    wt.onstatechange = (e) => console.log(e);
    await wt.ready;
    const writer = wt.datagramWritable.getWriter();
    const data1 = new Uint8Array([65, 66, 67]);
    const data2 = new Uint8Array([68, 69, 70]);
    await writer.ready;
    await writer.write(data1);
    await writer.ready;
    await writer.write(data2);
    await writer.close();
    await wt.datagramReadable.pipeTo(
      new WritableStream({
        write(value) {
          console.log(value, writer.desiredSize);
        },
        close() {
          // this is never reached
          console.log('Stream closed.');
        },
      })
    );
    // this is never reached
    await wt.close();
  } catch (err) {
    throw err;
  }
})().catch((err) => {
  console.error(err.message);
});
```
```
(async () => {
  try {
    const wt = new WebTransport('https://quic.aiortc.org/wt');
    wt.onstatechange = (e) => console.log(e);
    await wt.ready;
    const writer = wt.datagramWritable.getWriter();
    const data1 = new Uint8Array([65, 66, 67]);
    const data2 = new Uint8Array([68, 69, 70]);
    await writer.ready;
    await writer.write(data1);
    await writer.ready;
    await writer.write(data2);
    await writer.close();
    const reader = wt.datagramReadable.getReader();
    while (true) {
      const { value, done } = await reader.read();
      console.log(value, `writer.desiredSize: ${writer.desiredSize}`);
      if (done) {
        // this is never reached
        console.log('Stream closed.', done);
      }
    }
    // this is never reached
    await wt.close();
  } catch (err) {
    throw err;
  }
})().catch((err) => {
  console.error(err.message);
});
```

both versions produce the same result:

1) On first run `value` and `desiredSize` will be logged;
2) `datagramReadable` is never `done` though `writer.close()` is called;
3) `WebTransport` is never closed because 2) is never done;
4) Reload the page and run the same code several times and observe the error message: 

`Connection lost.`

Per the specification https://www.w3.org/TR/webtransport/#protocol-mapping `writable.getWriter.close()` method is mapped to QUIC Protocol Action `send STREAM_FINAL`

<table class="data">
    <colgroup class="header">
     <col>
    </colgroup><colgroup>
     <col>
    </colgroup><thead>
     <tr>
      <th>Method
      </th><th>QUIC Protocol Action
    </th></tr></thead><tbody>
     <tr>
      <td><code class="idl"><a data-link-type="idl" href="#dom-receivestream-abortreading" id="ref-for-dom-receivestream-abortreading①">abortReading</a></code>
      </td><td>send STOP_SENDING with code
     </td></tr><tr>
      <td><code class="idl"><a data-link-type="idl" href="#dom-sendstream-abortwriting" id="ref-for-dom-sendstream-abortwriting①">abortWriting</a></code>
      </td><td>send RESET_STREAM with code
     </td></tr><tr>
      <td><code class="idl"><a data-link-type="idl" href="#dom-bidirectionalstream-writable" id="ref-for-dom-bidirectionalstream-writable">writable</a></code>.abort()
      </td><td>send RESET_STREAM
     </td></tr><tr>
      <td><code class="idl"><a data-link-type="idl" href="#dom-bidirectionalstream-writable" id="ref-for-dom-bidirectionalstream-writable①">writable</a></code>.close()
      </td><td>send STREAM_FINAL
     </td></tr><tr>
      <td><code class="idl"><a data-link-type="idl" href="#dom-bidirectionalstream-writable" id="ref-for-dom-bidirectionalstream-writable②">writable</a></code>.getWriter().write()
      </td><td>send STREAM
     </td></tr><tr>
      <td><code class="idl"><a data-link-type="idl" href="#dom-bidirectionalstream-writable" id="ref-for-dom-bidirectionalstream-writable③">writable</a></code>.getWriter().close()
      </td><td>send STREAM_FINAL
     </td></tr><tr>
      <td><code class="idl"><a data-link-type="idl" href="#dom-bidirectionalstream-writable" id="ref-for-dom-bidirectionalstream-writable④">writable</a></code>.getWriter().abort()
      </td><td>send RESET_STREAM
     </td></tr><tr>
      <td><code class="idl"><a data-link-type="idl" href="#dom-bidirectionalstream-readable" id="ref-for-dom-bidirectionalstream-readable">readable</a></code>.cancel()
      </td><td>send STOP_SENDING
     </td></tr><tr>
      <td><code class="idl"><a data-link-type="idl" href="#dom-bidirectionalstream-readable" id="ref-for-dom-bidirectionalstream-readable①">readable</a></code>.getReader().read()
      </td><td>receive STREAM or STREAM_FINAL
     </td></tr><tr>
      <td><code class="idl"><a data-link-type="idl" href="#dom-bidirectionalstream-readable" id="ref-for-dom-bidirectionalstream-readable②">readable</a></code>.getReader().cancel()
      </td><td>send STOP_SENDING
   </td></tr></tbody></table>

which does not appear to be occurring.

This occurs with the stream from `createBidirectionalStream()` as well

```
(async () => {
  try {
    let transport = new WebTransport('https://quic.aiortc.org/wt');
    await transport.ready;

    let stream = await transport.createBidirectionalStream();
    let reader = stream.readable.getReader();
    let writer = stream.writable.getWriter();

    await writer.write(new Uint8Array([65, 66, 67]));
    // close writer
    await writer.close();
    while (true) {
      const { value, done } = await reader.read();
      console.log(value, done);
      if (done) {
        // this is never reached
        console.log(value, done);
        break;
      }
    }
    // this is never reached
    await wt.close();
  } catch (err) {
    throw err;
  }
})().catch((err) => {
  console.error(err.message);
});
```

**Comment by jlaine** [comment]
The demo server is just the bare minimum to get things started, this is still very early days for WebTransport.

For bidirectional streams it could indeed close the server -> client  when the stream in the other direction is closed. For datagrams I don't even know if there is a way to shutdown other than completely aborting the WebTransport session.

Feel free to suggest a PR.

**Comment by guest271314** [comment]
I will re-read the specification to make sure what is expected. W3C banned me so I cannot request clarification in the webtransport GitHub repository. In other API's where Streams API is used when the writable side of a transform stream is closed that is a signal to the readable side to eventually produce `{value: undefined, done: true}`. That is rather important when streaming media and need to be certain all of the bytes have been written, to avoid clipping at the end of the audio output - in this case there is no real way to know the stream is done at all. Observing the output at terminal I did not notice `STREAM_FINAL` ever being printed by the server.

**Comment by guest271314** [comment]
I was just doing some preliminary testing before updating the `WebTransport` sample at https://github.com/GoogleChrome/samples/blob/gh-pages/webtransport/quic_transport_server.py. I removed the `WebSocket` code from http3_server.py and was attempting to merge demo.py into http3_server.py - though the separate file might be useful. Currently I cannot update that sample code with the aioquic http3_server.py as it is - because the stream never ends where it did when using `QuicTransport` with aioquic. That is the reason I asked for server code to test at the front-end. I advised WebTransport contributors that the proverbial cart was being put before the horse re deprecation of QuicTransport and quic-transport - before having HTTP/3 working code - and got banned from blink-dev for that, too, under the auspices of having an "agressive" "tone". Well, `QuicTransport` with `quic-transport` worked.
